haystack_experimental/components/generators/chat/openai.py

```diff
@@ -0,0 +1,426 @@
+
+    def _connect_chunks(self, chunk: Any, chunks: List[StreamingChunk]) -> ChatMessage:
+        """
+        Connects the streaming chunks into a single ChatMessage.
+
+        :param chunk: The last chunk returned by the OpenAI API.
+        :param chunks: The list of all chunks returned by the OpenAI API.
+        """
+
+        # if it's a tool call , we need to build the payload dict from all the chunks
+        if bool(chunks[0].meta.get("tool_calls")):
+            tools_len = len(chunks[0].meta.get("tool_calls", []))
+
+            payloads = [{"arguments": "", "name": ""} for _ in range(tools_len)]
+            for chunk_payload in chunks:
+                deltas = chunk_payload.meta.get("tool_calls") or []
+
+                # deltas is a list of ChoiceDeltaToolCall or ChoiceDeltaFunctionCall
+                for i, delta in enumerate(deltas):
+                    payloads[i]["id"] = delta.id or payloads[i].get("id", "")
+                    if delta.function:
+                        payloads[i]["name"] += delta.function.name or ""
+                        payloads[i]["arguments"] += delta.function.arguments or ""
+
+            # handle potential malformed JSON strings
+            tools_calls = [
+                ToolCall(id=payload["id"], tool_name=payload["name"], arguments=json.loads(payload["arguments"]))
+                for payload in payloads
+            ]
+            complete_response = ChatMessage.from_assistant(tool_calls=tools_calls)
+        else:
+            complete_response = ChatMessage.from_assistant("".join([chunk.content for chunk in chunks]))
+        complete_response.meta.update(
+            {
+                "model": chunk.model,
+                "index": 0,
+                "finish_reason": chunk.choices[0].finish_reason,
+                "usage": {},  # we don't have usage data for streaming responses
+            }
+        )
+        return complete_response
+
+    def _build_message(self, completion: ChatCompletion, choice: Choice) -> ChatMessage:
+        """
+        Converts the non-streaming response from the OpenAI API to a ChatMessage.
+
+        :param completion: The completion returned by the OpenAI API.
+        :param choice: The choice returned by the OpenAI API.
+        :return: The ChatMessage.
+        """
+        message: ChatCompletionMessage = choice.message
+        text = message.content or ""
+        tool_calls = []
+        if openai_tool_calls := message.tool_calls:
+            for openai_tc in openai_tool_calls:
+                arguments_str = openai_tc.function.arguments
+                try:
+                    arguments = json.loads(arguments_str)
+                except json.JSONDecodeError:
+                    arguments = {"_malformed_json": arguments_str}
+                    logger.warning(
+                        "OpenAI returned a malformed JSON string. "
+                        "You can find it under the `_malformed_json` key of `ToolCall.arguments`."
```

Comment:
If it's malformed, then we just bail and (re-)raise with some extra context - I don't see how we can handle such exceptions.

Reply:
We can start with an approach similar to what you suggest, then see if users ask to return the invalid JSON string...

Since there may be more than one tool call and some of them may be valid, I would prefer to skip the invalid ones and emit a clear warning (instead of raising an exception). I will try this approach.